pkg/apimachinery/announced/group_factory.go

```diff
@@ -0,0 +1,244 @@
+// programmer importing the wrong set of packages. If this assumption doesn't
+// work for you, just call DefaultAnnouncementRegistry.AnnouncePreconstructedFactory
+// yourself.
+func (gmf *GroupMetaFactory) Announce() *GroupMetaFactory {
+	if err := DefaultAnnouncementRegistry.AnnouncePreconstructedFactory(gmf); err != nil {
+		panic(err)
+	}
+	return gmf
+}
+
+// GroupMetaFactory has the logic for actually assembling and registering a group.
+//
+// There are two ways of obtaining one of these.
+// 1. You can announce your group and versions separately, and then let the
+//    GroupFactoryRegistry assemble this object for you. (This allows group and
+//    versions to be imported separately, without referencing each other, to
+//    keep import trees small.)
+// 2. You can call NewGroupMetaFactory(), which is mostly a drop-in replacement
+//    for the old, bad way of doing things. You can then call .Announce() to
+//    announce your constructed factory to any code that would like to do
+//    things the new, better way.
+type GroupMetaFactory struct {
+	GroupArgs *GroupMetaFactoryArgs
+	// map of version name to version factory
+	VersionArgs map[string]*GroupVersionFactoryArgs
+
+	// assembled by Register()
+	prioritizedVersionList []unversioned.GroupVersion
+}
+
+// Register constructs the finalized prioritized version list and sanity checks
+// the announced group & versions. Then it calls register.
+func (gmf *GroupMetaFactory) Register(m *registered.APIRegistrationManager) error {
+	if gmf.GroupArgs == nil {
+		return fmt.Errorf("partially announced groups are not allowed, only got versions: %#v", gmf.VersionArgs)
+	}
+	if len(gmf.VersionArgs) == 0 {
+		return fmt.Errorf("group %v announced but no versions announced", gmf.GroupArgs.GroupName)
+	}
+
+	pvSet := sets.NewString(gmf.GroupArgs.VersionPreferenceOrder...)
+	if pvSet.Len() != len(gmf.GroupArgs.VersionPreferenceOrder) {
+		return fmt.Errorf("preference order for group %v has duplicates: %v", gmf.GroupArgs.GroupName, gmf.GroupArgs.VersionPreferenceOrder)
+	}
+	prioritizedVersions := []unversioned.GroupVersion{}
+	for _, v := range gmf.GroupArgs.VersionPreferenceOrder {
+		prioritizedVersions = append(
+			prioritizedVersions,
+			unversioned.GroupVersion{
+				gmf.GroupArgs.GroupName,
+				v,
+			},
+		)
+	}
+
+	availableVersions := []unversioned.GroupVersion{}
+	for _, v := range gmf.VersionArgs {
+		if v.GroupName != gmf.GroupArgs.GroupName {
+			return fmt.Errorf("found %v/%v in group %v?", v.GroupName, v.VersionName, gmf.GroupArgs.GroupName)
+		}
+		if pvSet.Has(v.VersionName) {
+			pvSet.Delete(v.VersionName)
+			continue
+		}
+		availableVersions = append(availableVersions, unversioned.GroupVersion{v.GroupName, v.VersionName})
+	}
+	if len(availableVersions) != 0 {
+		glog.Warningf("group %v has unprioritized versions %#v. They will have an arbitrary preference order!", gmf.GroupArgs.GroupName, availableVersions)
```

Comment:
Actually it's only problematic if there's more than one of these, now that I think about it :)